osfoffline/utils/authentication.py

```diff
@@ -0,0 +1,196 @@
+import asyncio
+import datetime
+import furl
+import logging
+
+import aiohttp
+import bcrypt
+from PyQt5.QtWidgets import QMessageBox
+
+from osfoffline import settings
+from osfoffline.database_manager.db import session
+from osfoffline.database_manager.models import User
+from osfoffline.database_manager.utils import save
+from osfoffline.polling_osf_manager.remote_objects import RemoteUser
+
+
+def generate_hash(password):
+    """ Generates a password hash using `bcrypt`.
+        Number of rounds for salt generation is 12.
+
+    :return: hashed password
+    """
+    return bcrypt.hashpw(
+        password.encode('utf-8'),
+        bcrypt.gensalt(12)
+    )
+
+token_request_body = json.loads(
+    '{"data": {"type": "tokens", "attributes": {"name": "OSF-Offline {}", "scopes": "osf.full_write"}}}'.format(datetime.date.today())
+)
+
+
+class AuthClient(object):
+    """Manages authorization flow """
+    def __init__(self):
+        self.API_URL = furl.furl(settings.API_BASE)
+        self.failed_login = False
+
+    @asyncio.coroutine
+    def _authenticate(self, username, password):
+        """ Tries to use standard auth to authenticate and create a personal access token through APIv2
+
+            :return: personal_access_token or None
+        """
+        token_url = self.API_URL.path.add('/v2/tokens/create/')
+        token_url.args['username'] = username
+        token_url.args['password'] = password
+
+        try:
+            resp = yield from aiohttp.request(method='POST', url=token_url.url, data=token_request_body)
+        except (aiohttp.errors.ClientTimeoutError, aiohttp.errors.ClientConnectionError, aiohttp.errors.TimeoutError):
+            # No internet connection
+            QMessageBox.warning(
+                None,
+                "Log in Failed",
+                "Unable to connect to server. Check your internet connection or try again later."
+            )
+            self.failed_login = True
+        except Exception as e:
+            # Invalid credentials
+            QMessageBox.warning(
+                None,
+                "Log in Failed",
+                "Invalid login credentials."
+            )
+            self.failed_login = True
+        else:
+            if not resp.status == 201:
+                return None
+            json_resp = yield from resp.json()
+            return json_resp['data']['attributes']['token_id']
+
+    @asyncio.coroutine
+    def _find_or_create_user(self, username, password):
+        """ Checks to see if username exists in DB and compares password hashes.
+            Tries to authenticate and create user if a valid user/oauth_token is not found.
+
+        :return: user
+        """
+        try:
+            user = session.query(User).filter(User.osf_login == username).one()
+            assert user.oauth_token
+            assert user.password = generate_hash(password)
```

Comment:
In general, assertions should be for things that you never ever expect to happen- try to avoid using them for testing regular error conditions.

Also, this try/except block is checking for two different and unrelated kinds of errors; if possible try to keep as little stuff as possible in a given try block, and only test for one error condition at a time. (bad db query vs wrong password)